src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs

```diff
@@ -65,51 +70,69 @@ internal partial class DiagnosticService : IDiagnosticService
+                if (!argsCollection.IsEmpty)
                 {
-                    _eventListenerTracker.EnsureEventListener(args.Workspace, this);
-                    previousWorkspace = args.Workspace;
+                    var updatedArgsCollection = UpdateDataMap(source, argsCollection);
+                    if (updatedArgsCollection.IsEmpty)
+                    {
+                        // there is no change, nothing to raise events for.
+                        continue;
+                    }
+
+                    ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.updatedArgsCollection), (source, updatedArgsCollection));
                 }
-            }
+                else
+                {
+                    using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
 
-            var ev = _eventMap.GetEventHandlers<EventHandler<ImmutableArray<DiagnosticsUpdatedArgs>>>(DiagnosticsUpdatedEventName);
+                    if (!ClearDiagnosticsReportedBySource(source, ref argsBuilder.AsRef()))
+                    {
+                        // there is no change, nothing to raise events for.
+                        continue;
+                    }
 
-            _eventQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
-            {
-                var updatedArgsCollection = UpdateDataMap(source, argsCollection);
-                if (updatedArgsCollection.IsEmpty)
-                {
-                    // there is no change, nothing to raise events for.
-                    return;
+                    // don't create event listener if it haven't created yet. if there is a diagnostic to remove
+                    // listener should have already created since all events are done in the serialized queue
+                    ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.args), (source, args: argsBuilder.ToImmutableAndClear()));
                 }
+            }
 
-                ev.RaiseEvent(static (handler, arg) => handler(arg.source, arg.updatedArgsCollection), (source, updatedArgsCollection));
-            }, CancellationToken.None);
+            return ValueTaskFactory.CompletedTask;
         }
 
-        private void RaiseDiagnosticsCleared(IDiagnosticUpdateSource source)
+        private void RaiseDiagnosticsUpdated(IDiagnosticUpdateSource source, ImmutableArray<DiagnosticsUpdatedArgs> argsCollection)
         {
-            var ev = _eventMap.GetEventHandlers<EventHandler<ImmutableArray<DiagnosticsUpdatedArgs>>>(DiagnosticsUpdatedEventName);
-
-            _eventQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
+            if (argsCollection.IsEmpty)
             {
-                using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
+                // No work to do. Avoid adding an empty args collection to _eventQueue because that is used to represent
+                // an operation from RaiseDiagnosticsCleared.
```

Comment:
is this the same logic we had preivously?  trying to find it.

Reply:
ok. i think i've wrapped my head around this.  but i def find it confusing.  could we instead enqueue this information as a boolean into the ABWQ instead?

Reply:
⏱️ Switching to an enum

Reply:
➡️ Updated